crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1900,12 +1900,11 @@ impl<'db> TypeInferenceBuilder<'db> {
             is_async: _,
         } = for_statement;
 
-        self.infer_standalone_expression(iter);
-
         // TODO more complex assignment targets
         if let ast::Expr::Name(name) = &**target {
             self.infer_definition(name);
         } else {
+            self.infer_standalone_expression(iter);
```

Comment:
Why is it okay to only infer the iterator when the target isn't a name? Is it because it is otherwise handled in `infer_for_statement_definition` or is it something else?

Reply:
Yes, that's correct. It's handled while inferring the definition types. This pattern is also followed in other definitions like assignment, with items, etc.

Reply:
The main reason for this requirement is the `self.extend` call which duplicated the diagnostics.